ambry-tools/src/main/java/com.github.ambry/clustermap/HelixBootstrapUpgradeTool.java

```diff
@@ -0,0 +1,667 @@
+        InstanceConfig instanceConfig = dcAdmin.getInstanceConfig(clusterName, instanceName);
+        List<String> currentSealedPartitions = instanceConfig.getRecord().getListField(SEALED_STR);
+        List<String> newSealedPartitionsList = new ArrayList<>(currentSealedPartitions);
+        if (isSealed && !currentSealedPartitions.contains(partitionName)) {
+          newSealedPartitionsList.add(partitionName);
+        } else if (!isSealed && currentSealedPartitions.contains(partitionName)) {
+          newSealedPartitionsList.remove(partitionName);
+        }
+        instanceConfig.getRecord().setListField(SEALED_STR, newSealedPartitionsList);
+        // @todo: uncomment when we move to Helix 0.6.7.
+        // dcAdmin.setInstanceConfig(clusterName, instanceName, instanceConfig);
+      }
+    }
+  }
+
+  /**
+   * Adds all partitions to every datacenter with replicas in nodes as specified in the static clustermap (unless it
+   * was already added).
+   *
+   * The assumption is that in the static layout, every partition is contained in every colo. We make this assumption
+   * to ensure that partitions are grouped under the same resource in all colos (since the resource id is not
+   * something that is present today in the static cluster map). This is not a strict requirement though, but helps
+   * ease the logic.
+   *
+   * Note: 1. We ensure that the partition names are unique in the Ambry cluster even across resources.
+   *       2. New Ambry partitions will not be added to Helix resources that are already present before the call to this
+   *          method.
+   */
+  private void addNewAmbryPartitions(List<Partition> partitions, String resourceName) {
+    // In the future, a resource may be used to group together partitions of a container. For now, multiple
+    // resources are created and partitions are grouped under these resources upto a maximum threshold.
+    if (partitions.isEmpty()) {
+      throw new IllegalArgumentException("Cannot add resource with zero partitions");
+    }
+    for (Map.Entry<String, HelixAdmin> entry : adminForDc.entrySet()) {
+      String dcName = entry.getKey();
+      HelixAdmin dcAdmin = entry.getValue();
+      AutoModeISBuilder resourceISBuilder = new AutoModeISBuilder(resourceName);
+      int numReplicas = 0;
+      resourceISBuilder.setStateModel(LeaderStandbySMD.name);
+      for (Partition partition : partitions) {
+        String partitionName = Long.toString(partition.getId());
+        boolean sealed = partition.getPartitionState().equals(PartitionState.READ_ONLY);
+        List<ReplicaId> replicaList = getReplicasInDc(partition, dcName);
+        numReplicas = replicaList.size();
+        String[] instances = updateInstancesAndGetInstanceNames(dcAdmin, partitionName, replicaList, sealed);
+        Collections.shuffle(Arrays.asList(instances));
+        resourceISBuilder.assignPreferenceList(partitionName, instances);
+      }
+      resourceISBuilder.setNumReplica(numReplicas);
```

Comment:
Does resource have information about no of partitions it has ? Or just get the list of partitions and get the size of them ?